File: ZezeJava/ZezeJava/src/main/java/Zeze/Arch/ProviderDirectService.java
package Zeze.Arch;

import java.net.ServerSocket;
import java.util.concurrent.ConcurrentHashMap;
import Zeze.Application;
import Zeze.Builtin.Provider.BModule;
import Zeze.Builtin.ProviderDirect.AnnounceProviderInfo;
import Zeze.Builtin.ProviderDirect.ModuleRedirect;
import Zeze.Builtin.ProviderDirect.ModuleRedirectAllResult;
import Zeze.Net.AsyncSocket;
import Zeze.Net.Connector;
import Zeze.Net.Protocol;
import Zeze.Net.ProtocolHandle;
import Zeze.Serialize.ByteBuffer;
import Zeze.Services.HandshakeBoth;
import Zeze.Services.ServiceManager.BServiceInfo;
import Zeze.Util.Action0;
import Zeze.Util.ConcurrentHashSet;
import Zeze.Util.LongConcurrentHashMap;
import Zeze.Util.OutObject;
import Zeze.Util.Task;
import Zeze.Util.TaskCompletionSource;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Provider之间直连网络管理服务。
 */
public class ProviderDirectService extends HandshakeBoth {
	private static final @NotNull Logger logger = LogManager.getLogger(ProviderDirectService.class);

	protected ProviderApp providerApp;
	public final ConcurrentHashMap<String, ProviderSession> providerByLoadName = new ConcurrentHashMap<>();
	public final LongConcurrentHashMap<ProviderSession> providerByServerId = new LongConcurrentHashMap<>();
	private final LongConcurrentHashMap<ConcurrentHashSet<Action0>> serverReadyEvents = new LongConcurrentHashMap<>();

	public ProviderDirectService(@NotNull String name, @NotNull Application zeze) {
		super(name, zeze);
	}

	public void removeServer(@NotNull BServiceInfo pm) {
		lock();
		try {
			var ss = providerApp.zeze.getServiceManager().getSubscribeStates().get(pm.getServiceName());
			if (null != ss) {
				var connName = pm.getPassiveIp() + "_" + pm.getPassivePort();
				var conn = getConfig().findConnector(connName);
				if (conn != null) {
					conn.stop();
					providerByLoadName.remove(connName);
					var serverId = Integer.parseInt(pm.getServiceIdentity());
					providerByServerId.remove(serverId);
					ss.setIdentityLocalState(pm.getServiceIdentity(), null);
					getConfig().removeConnector(conn);
				}
			}
		} finally {
			unlock();
		}
	}

	public void addServer(@NotNull BServiceInfo pm) {
		lock();
		try {
			var connName = pm.getPassiveIp() + "_" + pm.getPassivePort();
			var ps = providerByLoadName.get(connName);
			if (ps != null) {
				// connection has ready.
				var mid = Integer.parseInt(pm.getServiceName().split("#")[1]);
				var m = providerApp.modules.get(mid);
				if (m != null)
					setReady(pm.getServiceName(), pm, ps, mid, m);
				else
					logger.error("addServer: not found module: {}", pm.getServiceName());
				return;
			}
			if (pm.getServiceIdentity().startsWith("@")) // from linkd
				return;
			var serverId = Integer.parseInt(pm.getServiceIdentity());
			if (serverId < getZeze().getConfig().getServerId())
				return;
			if (serverId == getZeze().getConfig().getServerId()) {
				setRelativeServiceReady(newSession(serverId), providerApp.directIp, providerApp.directPort);
				return;
			}
			var out = new OutObject<Connector>();
			if (getConfig().tryGetOrAddConnector(pm.getPassiveIp(), pm.getPassivePort(), true, out)) {
				// 新建的Connector。开始连接。
				out.value.userState = newSession(serverId);
				out.value.start();
			}
		} finally {
			unlock();
		}
	}

	@SuppressWarnings("MethodMayBeStatic")
	public @NotNull ProviderSession newSession(int serverId) {
		var session = new ProviderSession();
		session.serverId = serverId;
		return session;
	}

	@SuppressWarnings("MethodMayBeStatic")
	public @NotNull ProviderSession newSession(@NotNull AsyncSocket so) {
		var session = new ProviderSession();
		session.sessionId = so.getSessionId();
		return session;
	}

	@Override
	public void OnSocketAccept(@NotNull AsyncSocket so) {
		if (so.getConnector() == null) {
			// 被动连接等待对方报告信息时再处理。
			// passive connection continue process in ProviderDirect.ProcessAnnounceProviderInfoRequest.
			so.setUserState(newSession(so)); // acceptor
		}
		super.OnSocketAccept(so);
	}

	@Override
	public void OnHandshakeDone(@NotNull AsyncSocket so) throws Exception {
		// call base
		super.OnHandshakeDone(so);

		var c = so.getConnector();
		if (c != null) {
			// 主动连接。
			var ps = (ProviderSession)so.getUserState();
			ps.sessionId = so.getSessionId();
			setRelativeServiceReady(ps, c.getHostNameOrAddress(), c.getPort());

			var r = new AnnounceProviderInfo();
			r.Argument.setIp(providerApp.directIp);
			r.Argument.setPort(providerApp.directPort);
			r.Argument.setServerId(providerApp.zeze.getConfig().getServerId());
			r.Send(so, (_r) -> 0L); // skip result
		}
	}

	public void waitDirectServerReady(int serverId) {
		waitDirectServerReady(serverId, 3000);
	}

	public void waitDirectServerReady(int serverId, long timeout) {
		var future = new TaskCompletionSource<Long>();
		Action0 callback = () -> future.setResult(0L);
		try {
			waitDirectServerReady(serverId, callback);
			future.await(timeout);
		} finally {
			serverReadyEvents.computeIfAbsent(serverId, __ -> new ConcurrentHashSet<>()).remove(callback);
		}
	}

	public void waitDirectServerReady(int serverId, @NotNull Action0 callback) {
		lock();
		try {
			if (!providerByServerId.containsKey(serverId)) {
				serverReadyEvents.computeIfAbsent(serverId, __ -> new ConcurrentHashSet<>()).add(callback);
				return;
			}
		} finally {
			unlock();
		}
		try {
			callback.run(); // 锁外回调，避免死锁风险。
		} catch (Exception e) {
			Task.forceThrow(e);
		}
	}

	// 由于sm的服务信息是碎片传递给订阅者的，所以本质上得到的快照在启动的时候几乎总是不完整的，
	// 为了得到真正所有的服务器信息，只有sleep。
	// 先不提供waitAllDirectServerReady了。
	/*
	public void waitAllDirectServerReady(Action0 callback) {
		// 得到当前provider服务集合（快照）。
		var servers = new HashSet<Integer>();
		for (var ss : providerApp.zeze.getServiceManager().getSubscribeStates().values()) {
			if (ss.getServiceName().startsWith(providerApp.serverServiceNamePrefix)) {
				for (var info : ss.getServiceInfos().getServiceInfoListSortedByIdentity()) {
					var serverId = Integer.parseInt(info.serviceIdentity);
					servers.add(serverId);
				}
			}
		}
		// 得到没有direct没有好的。
		var pending = new HashSet<Integer>();
		lock();
		try {
			for (var serverId : servers) {
				if (!providerByServerId.containsKey(serverId))
					pending.add(serverId);
			}
		} finally {
			unlock();
		}
		// 订阅没有准备好的。
		for (var pend : pending) {

		}
	}
	*/

	// under lock
	private void notifyServerReady(int serverId) {
		var watchers = serverReadyEvents.computeIfAbsent(serverId, __ -> new ConcurrentHashSet<>());
		for (var w : watchers) {
			try {
				w.run();
			} catch (Exception ex) {
				logger.error("", ex);
			}
		}
		watchers.clear();
	}

	void setRelativeServiceReady(@NotNull ProviderSession ps, @NotNull String ip, int port) {
		lock();
		try {
			ps.serverLoadIp = ip;
			ps.serverLoadPort = port;
			// 本机的连接可能设置多次。此时使用已经存在的，忽略后面的。
			if (providerByLoadName.putIfAbsent(ps.getServerLoadName(), ps) != null)
				return;
			providerByServerId.put(ps.getServerId(), ps);

			// 需要把所有符合当前连接目标的Provider相关的服务信息都更新到当前连接的状态。
			for (var ss : getZeze().getServiceManager().getSubscribeStates().values()) {
				if (ss.getServiceName().startsWith(providerApp.serverServiceNamePrefix)) {
					var infos = ss.getServiceInfos(ps.appVersion);
					if (infos == null)
						continue;
					var mid = Integer.parseInt(ss.getServiceName().split("#")[1]);
					var m = providerApp.modules.get(mid);
					if (m == null) {
						logger.error("setRelativeServiceReady: not found module: {}", ss.getServiceName());
						continue;
					}
					for (var server : infos.getSortedIdentities()) {
						// 符合当前连接目标。每个Identity标识的服务的(ip,port)必须不一样。
						if (server.getPassiveIp().equals(ip) && server.getPassivePort() == port) {
							setReady(ss.getServiceName(), server, ps, mid, m);
						}
					}
				}
			}
			// 最后才通知成功。
			notifyServerReady(ps.getServerId());
		} finally {
			unlock();
		}
	}

	private void setReady(@NotNull String serviceName, @NotNull BServiceInfo server, @NotNull ProviderSession ps,
						  int mid, @NotNull BModule.Data m) {
		var ss = providerApp.zeze.getServiceManager().getSubscribeStates().get(serviceName);
		if (null != ss) {
			var pms = new ProviderModuleState(ps.getSessionId(), mid, m.getChoiceType(), m.getConfigType());
			ps.getOrAddServiceReadyState(serviceName).put(server.getServiceIdentity(), pms);
			ss.setIdentityLocalState(server.getServiceIdentity(), pms);
		}
	}

	@Override
	public void OnSocketClose(@NotNull AsyncSocket socket, @Nullable Throwable ex) throws Exception {
		var ps = (ProviderSession)socket.getUserState();
		if (ps != null) {
			for (var service : ps.ServiceReadyStates.entrySet()) {
				var subs = getZeze().getServiceManager().getSubscribeStates().get(service.getKey());
				for (var identity : service.getValue().keySet()) {
					subs.setIdentityLocalState(identity, null);
				}
			}
			providerByLoadName.remove(ps.getServerLoadName());
			providerByServerId.remove(ps.getServerId());
		}
		super.OnSocketClose(socket, ex);
	}

	@Override
	public void dispatchProtocol(long typeId, @NotNull ByteBuffer bb, @NotNull ProtocolFactoryHandle<?> factoryHandle,
								 @Nullable AsyncSocket so) throws Exception {
		var p = decodeProtocol(typeId, bb, factoryHandle, so);
		p.dispatch(this, factoryHandle);
	}

	@Override
	public void dispatchProtocol(@NotNull Protocol<?> p, @NotNull ProtocolFactoryHandle<?> factoryHandle)
			throws Exception {
		if (p.getTypeId() == ModuleRedirect.TypeId_) {
			var r = (ModuleRedirect)p;
			// 总是不启用存储过程，内部处理redirect时根据Redirect.Handle配置决定是否在存储过程中执行。
			if (r.Argument.isNoOneByOne()) {
				Task.executeUnsafe(() -> p.handle(this, factoryHandle), p, Protocol::trySendResultCode,
						r.Argument.getMethodFullName(), factoryHandle.Mode);
			} else {
				getZeze().getTaskOneByOneByKey().Execute(r.Argument.getKey(),
						() -> Task.call(() -> p.handle(this, factoryHandle), p,
								Protocol::trySendResultCode, r.Argument.getMethodFullName()),
						factoryHandle.Mode);
			}
			return;
		}
		if (p.getTypeId() == ModuleRedirectAllResult.TypeId_) {
			var r = (ModuleRedirectAllResult)p;
			// 总是不启用存储过程，内部处理redirect时根据Redirect.Handle配置决定是否在存储过程中执行。
			Task.executeUnsafe(() -> p.handle(this, factoryHandle), p, Protocol::trySendResultCode,
					r.Argument.getMethodFullName(), factoryHandle.Mode);
			return;
		}
		// 所有的Direct都不启用存储过程。
		Task.executeUnsafe(() -> p.handle(this, factoryHandle), p, Protocol::trySendResultCode, null,
				factoryHandle.Mode);
		//super.DispatchProtocol(p, factoryHandle);
	}

	@Override
	public <P extends Protocol<?>> void dispatchRpcResponse(@NotNull P rpc, @NotNull ProtocolHandle<P> responseHandle,
															@NotNull ProtocolFactoryHandle<?> factoryHandle) {
		if (rpc.getTypeId() == ModuleRedirect.TypeId_) {
			var r = (ModuleRedirect)rpc;
			// 总是不启用存储过程，内部处理redirect时根据Redirect.Handle配置决定是否在存储过程中执行。
			if (r.Argument.isNoOneByOne())
				Task.executeRpcResponseUnsafe(() -> responseHandle.handle(rpc), rpc, factoryHandle.Mode);
			else {
				getZeze().getTaskOneByOneByKey().Execute(r.Argument.getKey(),
						() -> Task.call(() -> responseHandle.handle(rpc), rpc), factoryHandle.Mode);
			}
			return;
		}

		// no procedure.
		Task.executeRpcResponseUnsafe(() -> responseHandle.handle(rpc), rpc, factoryHandle.Mode);
		//super.dispatchRpcResponse(rpc, responseHandle, factoryHandle);
	}

	@Override
	public void onServerSocketBind(@NotNull ServerSocket ss) {
		providerApp.directPort = ss.getLocalPort();
	}
}

File: Zeze/Arch/ProviderDirectService.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;
using Zeze.Builtin.ProviderDirect;
using Zeze.Net;
using Zeze.Services.ServiceManager;
using Zeze.Util;

namespace Zeze.Arch
{
	public class ProviderDirectService : Zeze.Services.HandshakeBoth
	{
		private static readonly new ILogger logger = LogManager.GetLogger(typeof(ProviderDirectService));

		public ProviderApp ProviderApp;
		public readonly ConcurrentDictionary<string, ProviderSession> ProviderByLoadName = new();
		public readonly ConcurrentDictionary<int, ProviderSession> ProviderByServerId = new();
		public readonly ConcurrentDictionary<int, IdentityHashSet<Action>> ServerReadyEvents = new();

        public ProviderDirectService(string name, Zeze.Application zeze)
			: base(name, zeze)
		{
		}

		public void RemoveServer(ServiceInfo pm)
		{
            if (false == ProviderApp.Zeze.ServiceManager.SubscribeStates.TryGetValue(pm.ServiceIdentity, out var ss))
                return;

            lock (this)
			{
				var connName = pm.PassiveIp + ":" + pm.PassivePort;
				var conn = Config.FindConnector(connName);
				if (null != conn)
				{
					conn.Stop();
					ProviderByLoadName.TryRemove(connName, out _);
					ProviderByServerId.TryRemove(int.Parse(pm.ServiceIdentity), out _);
					ss.SetIdentityLocalState(pm.ServiceIdentity, null);
					Config.RemoveConnector(conn);
				}
			}
		}

		public void AddServer(ServiceInfo pm)
		{
			if (false == ProviderApp.Zeze.ServiceManager.SubscribeStates.TryGetValue(pm.ServiceIdentity, out var ss))
				return;

			lock (this)
			{
				var connName = pm.PassiveIp + ":" + pm.PassivePort;
				if (ProviderByLoadName.TryGetValue(connName, out var ps))
				{
					// connection has ready.
					var mid = int.Parse(pm.ServiceName.Split('#')[1]);
					if (false == ProviderApp.Modules.TryGetValue(mid, out var m))
						throw new Exception($"Module Not Found {mid}");
					SetReady(ss, pm, ps, mid, m);
					return;
				}
				var serverId = int.Parse(pm.ServiceIdentity);
				if (serverId < Zeze.Config.ServerId)
					return;
				if (serverId == Zeze.Config.ServerId)
				{
					var psLocal = new ProviderSession
					{
						ServerId = serverId
					};
					SetRelativeServiceReady(psLocal, ProviderApp.DirectIp, ProviderApp.DirectPort);
					return;
				}
				if (Config.TryGetOrAddConnector(pm.PassiveIp, pm.PassivePort, true, out var newAdd))
				{
					// 新建的Connector。开始连接。
					var psPeer = new ProviderSession
					{
						ServerId = serverId
					};
					newAdd.UserState = psPeer;
					newAdd.Start();
				}
			}
		}

        public override void OnSocketAccept(AsyncSocket socket)
        {
			if (socket.Connector == null)
            {
                // 被动连接等待对方报告信息时再处理。
                var ps = new ProviderSession
                {
                    SessionId = socket.SessionId
                };
                socket.UserState = ps;
			}
			base.OnSocketAccept(socket);
        }

        public override void OnHandshakeDone(AsyncSocket socket)
		{
			// call base
			base.OnHandshakeDone(socket);

			var c = socket.Connector;
			if (c != null) {
				// 主动连接。
				var ps = (ProviderSession)socket.UserState;
				ps.SessionId = socket.SessionId;
				SetRelativeServiceReady(ps, c.HostNameOrAddress, c.Port);
				var r = new AnnounceProviderInfo();
				r.Argument.Ip = ProviderApp.DirectIp;
				r.Argument.Port = ProviderApp.DirectPort;
				r.Send(socket, (_r) => Task.FromResult(0L)); // skip result
			}
		}

        public void WaitDirectServerReady(int serverId, int timeout = 3000)
        {
            var future = new TaskCompletionSource<int>();
            Action callback = () => future.SetResult(0);
            try
            {
                WaitDirectServerReady(serverId, callback);
                future.Task.Wait(timeout);
            }
            finally
            {
                ServerReadyEvents.GetOrAdd(serverId, _ => new IdentityHashSet<Action>()).Remove(callback);
            }
        }

        public void WaitDirectServerReady(int serverId, Action callback)
        {
            lock(this)
			{
                if (!ProviderByServerId.ContainsKey(serverId))
                {
                    ServerReadyEvents.GetOrAdd(serverId, _ => new IdentityHashSet<Action>()).Add(callback);
                    return;
                }
            }
            callback(); // 锁外回调，避免死锁风险。
        }

        // under lock
        private void NotifyServerReady(int serverId)
        {
            var watchers = ServerReadyEvents.GetOrAdd(serverId, _ => new IdentityHashSet<Action>());
            foreach (var w in watchers)
            {
                try
                {
                    w();
                }
                catch (Exception ex)
				{
                    logger.Error(ex);
                }
            }
            watchers.Clear();
        }
        
		internal void SetRelativeServiceReady(ProviderSession ps, String ip, int port)
		{
			lock (this)
			{
				ps.ServerLoadIp = ip;
				ps.ServerLoadPort = port;
				// 本机的连接可能设置多次。此时使用已经存在的，忽略后面的。
				if (false == ProviderByLoadName.TryAdd(ps.ServerLoadName, ps))
					return;
				ProviderByServerId[ps.ServerId] = ps;

				// 需要把所有符合当前连接目标的Provider相关的服务信息都更新到当前连接的状态。
				foreach (var ss in Zeze.ServiceManager.SubscribeStates.Values)
				{
					if (ss.ServiceName.StartsWith(ProviderApp.ServerServiceNamePrefix))
					{
						if (false == ss.ServiceInfosVersion.InfosVersion.TryGetValue(0, out var infos))
							continue;

						var mid = int.Parse(ss.ServiceName.Split('#')[1]);
						if (false == ProviderApp.Modules.TryGetValue(mid, out var m))
							throw new Exception($"Module Not Found {mid}");
						foreach (var server in infos.SortedIdentity)
						{
							// 符合当前连接目标。每个Identity标识的服务的(ip,port)必须不一样。
							if (server.PassiveIp.Equals(ip) && server.PassivePort == port)
							{
								SetReady(ss, server, ps, mid, m);
							}
						}
					}
				}
				NotifyServerReady(ps.ServerId);
            }
		}

		private void SetReady(Agent.SubscribeState ss, ServiceInfo server, ProviderSession ps,
			int mid, Zeze.Builtin.Provider.BModule m)
		{
			Console.WriteLine($"SetReady Server={Zeze.Config.ServerId} {ss.ServiceName} {server.ServiceIdentity}");
			var pms = new ProviderModuleState(ps.SessionId, mid, m.ChoiceType, m.ConfigType);
			ps.GetOrAddServiceReadyState(ss.ServiceName).TryAdd(server.ServiceIdentity, pms);
			ss.SetIdentityLocalState(server.ServiceIdentity, pms);
		}

		public override void OnSocketClose(AsyncSocket socket, Exception ex)
		{
			var ps = (ProviderSession)socket.UserState;
			if (ps != null)
			{
				foreach (var service in ps.ServiceReadyStates)
				{
					if (Zeze.ServiceManager.SubscribeStates.TryGetValue(service.Key, out var subs))
                    {
						foreach (var identity in service.Value.Keys)
						{
							subs.SetIdentityLocalState(identity, null);
						}
					}
				}
				ProviderByLoadName.TryRemove(ps.ServerLoadName, out _);
				ProviderByServerId.TryRemove(ps.ServerId, out _);
			}
			base.OnSocketClose(socket, ex);
		}

		public override void DispatchProtocol(Protocol p, ProtocolFactoryHandle factoryHandle)
		{
			if (p.TypeId == ModuleRedirect.TypeId_)
			{
				var r = (ModuleRedirect)p;
				// 总是不启用存储过程，内部处理redirect时根据Redirect.Handle配置决定是否在存储过程中执行。
				Zeze.TaskOneByOneByKey.Execute(
					r.Argument.HashCode, factoryHandle.Handle, p, r.Argument.MethodFullName,
					(p, code) => p.TrySendResultCode(code));

				return;
			}

			if (p.TypeId == ModuleRedirectAllResult.TypeId_)
			{
				var r = (ModuleRedirectAllResult)p;
				// 总是不启用存储过程，内部处理redirect时根据Redirect.Handle配置决定是否在存储过程中执行。
				_ = Mission.CallAsync(factoryHandle.Handle, p, (p, code) => p.TrySendResultCode(code), r.Argument.MethodFullName);

				return;
			}
			// 所有的ProviderDirectService都不启用存储过程。
			// 按收到顺序处理，不并发。这样也避免线程切换。
			_ = Mission.CallAsync(factoryHandle.Handle, p, (p, code) => p.TrySendResultCode(code));
		}

		public override void DispatchRpcResponse(Protocol rpc, Func<Protocol, Task<long>> responseHandle, ProtocolFactoryHandle factoryHandle)
		{
			if (rpc.TypeId == ModuleRedirect.TypeId_)
			{
				var redirect = (ModuleRedirect)rpc;
				// 总是不启用存储过程，内部处理redirect时根据Redirect.Handle配置决定是否在存储过程中执行。
				Zeze.TaskOneByOneByKey.Execute(redirect.Argument.HashCode, responseHandle, rpc);
				return;
			}

			// no procedure.
			// 按收到顺序处理，不并发。这样也避免线程切换。
			_ = Mission.CallAsync(responseHandle, rpc);
		}
	}
}
